src/vision/status_quo/alan_tries_a_socket_sink.md

```diff
@@ -0,0 +1,115 @@
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+Alan is working on a project that uses `async-std`. He has worked a bit with `tokio` in the past and is more familiar with that, but he is interested to learn something how things work in `async-std`.
+
+One of the goals is to switch from a WebSocket implementation using raw TCP sockets to one managed behind an HTTP server library, so both HTTP and WebSocket commands can be forwarded to a transport-agnostic RPC server. He finds the HTTP server `tide` and it seems fairly similar to `warp`, which he was using with `tokio`. He also finds the WebSocket middleware library `tide-websockets` that goes with it.
+
+However, as he's working, Alan encounters a situation where the socket needs to be written to within an async thread, and the traits just aren't working. He wants to split the stream into a sender and receiver:
+
+```rust
+use futures::{SinkExt, StreamExt};
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = ws_stream.split();
+let ws_sender = Arc::new(Mutex::new(ws_sender));
+
+while let Some(msg) = ws_receiver.next().await {
+    debug!("Received new WS RPC message: {:?}", msg);
+
+    async_std::task::spawn(async move {
+        let res = call_rpc(msg).await?;
+
+        match ws_sender.lock().await.send_string(res).await {
+            Ok(_) => info!("New WS data sent."),
+            Err(_) => warn!("WS connection closed."),
+        };
+    });
+}
+```
+
+This is necessary because both sides of the stream are stateful iterators, and if they're kept as one, the lock on the receiver can't be released within the loop. He's seen this pattern used in other projects in the Rust community, but is frustrated to find that the `Sink` trait wasn't implemented in the WebSockets middleware library he's using.
+
+Alan also tries creating a sort of poller worker thread using an intermediary messaging channel, but he has trouble reasoning about the code and wasn't able to get it to compile:
+
+```rust
+use async_std::channel;
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = channel::unbounded::<String>();
+let ws_receiver = Arc::new(ws_receiver);
+
+let ws_stream = Arc::new(Mutex::new(ws_stream));
+let poller_ws_stream = ws_stream.clone();
+
+async_std::task::spawn(async move {
```

Comment:
I'm starting to see the idea here, but it is indeed still confusing to me. this snippet of code doesn't seem to be starting with the same "general inputs". Maybe it'd be helpful if we could give a bit of an architectural diagram?

Reply:
The only input that matters (and is common between both examples) is `ws_stream`. The implementation details differ, but I think they're roughly equivalent in practice.